platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/serialize/DefaultGlobalValueCodec.kt

```diff
@@ -0,0 +1,187 @@
+
+    private
+    val globalContext by lazy(globalContextProvider)
+
+    private
+    inner class FutureValue {
+
+        private
+        val latch = CountDownLatch(1)
+
+        private
+        var value: Any? = null
+
+        fun complete(v: Any) {
+            value = v
+            latch.countDown()
+        }
+
+        fun get(): Any {
+            val state = state.get()
+            if (value == null && state < ReaderState.STOPPED && !latch.await(1, TimeUnit.MINUTES)) {
+                throw TimeoutException("Timeout while waiting for value, state was $state")
+            }
+            require(value != null) { "State is: $state"}
+            return value!!
+        }
+    }
+
+    private
+    val values = ConcurrentHashMap<Int, Any>()
+
+    /**
+     * Reads all values available (and their ids) to build the pool of shared objects
+     */
+    //TODO-RC Use a Gradle managed facility instead of an ad-hoc thread
+    private
+    val reader = thread(start = false, isDaemon = true, name = "${this::class.simpleName} reader thread") {
+        require(state.compareAndSet(ReaderState.STARTED, ReaderState.RUNNING)) {
+            "Unexpected state: $state"
+        }
+        globalContext.useToRun {
+            while (state.get() == ReaderState.RUNNING) {
+                val id = readSmallInt()
+                if (id == EOF) {
+                    stopReading()
+                    break
+                }
+                val read = runReadOperation {
+                    read()!!
+                }
+                values.compute(id) { _, value ->
+                    when (value) {
+                        is FutureValue -> value.complete(read)
+                        else -> require(value == null)
+                    }
+                    read
+                }
+            }
+            state.set(ReaderState.STOPPED)
+        }
+    }
+
+    private
+    fun ReadContext.resolveValue(id: Int): Any {
+        startReadingIfNeeded()
+        require(id >= 0) {
+            "id: $id - $this"
+        }
+        return when (val existing = values.computeIfAbsent(id) { FutureValue() }) {
+                is FutureValue -> existing.get()
+                else -> existing
+            }
+    }
+
+    private fun startReadingIfNeeded() {
+        if (state.compareAndSet(ReaderState.READY, ReaderState.STARTED)) {
+            reader.start()
+        }
+    }
+
+    override fun close() {
+        stopReading()
+        reader.join(TimeUnit.MINUTES.toMillis(1))
+    }
+
+    private fun stopReading() {
+        state.compareAndSet(ReaderState.RUNNING, ReaderState.STOPPING)
+    }
+
+    override suspend fun <T : Any> read(readContext: ReadContext, decode: suspend ReadContext.() -> T): T {
+        val id = readContext.readSmallInt()
+        return readContext.resolveValue(id).uncheckedCast()
+    }
+}
+
+fun <T: Any, C: IsolateContext> C.synchronized(action: C.() -> T?) = synchronized(this) {
```

Comment:
🤔 Is this adding any value?

Reply:
It ensures the object we are synchronizing on is the same we are accessing  in the action.

Now we can:

```
globalContext.synchronized {
    runWriteOperation {
        writeSmallInt(id)
        write(value)
    }
}
```  

instead of:

```
synchronized(globalContext) {
    globalContext.runWriteOperation {
        writeSmallInt(id)
        write(value)
    }
}
```